File: src/go/2_CreationalPatterns/1_Builder/functionalbuilder.go
package main

import "fmt"

type Person struct {
  name, position string
}

type personMod func(*Person)
type PersonBuilder struct {
  actions []personMod
}

func (b *PersonBuilder) Called(name string) *PersonBuilder {
  b.actions = append(b.actions, func(p *Person) {
    p.name = name
  })
  return b
}

func (b *PersonBuilder) Build() *Person {
  p := Person{}
  for _, a := range b.actions {
    a(&p)
  }
  return &p
}

// extend PersonBuilder
func (b *PersonBuilder) WorksAsA(position string) *PersonBuilder {
  b.actions = append(b.actions, func(p *Person) {
    p.position = position
  })
  return b
}

func main() {
  b := PersonBuilder{}
  p := b.Called("Dmitri").WorksAsA("dev").Build()
  fmt.Println(*p)
}

File: src/csharp/2_CreationalPatterns/1_Builder/FunctionalBuilder.cs
using System;
using System.Collections.Generic;

namespace DotNetDesignPatternDemos.Creational.Builder
{
  public class Person
  {
    public string Name, Position;
  }

  public sealed class PersonBuilder
  {
    public readonly List<Action<Person>> Actions 
      = new List<Action<Person>>();

    public PersonBuilder Called(string name)
    {
      Actions.Add(p => { p.Name = name; });
      return this;
    }

    public Person Build()
    {
      var p = new Person();
      Actions.ForEach(a => a(p));
      return p;
    }
  }

  public static class PersonBuilderExtensions
  {
    public static PersonBuilder WorksAsA
      (this PersonBuilder builder, string position)
    {
      builder.Actions.Add(p =>
      {
        p.Position = position;
      });
      return builder;
    }
  }
  
  public class FunctionalBuilder
  {
    public static void Main(string[] args)
    {
      var pb = new PersonBuilder();
      var person = pb.Called("Dmitri").WorksAsA("Programmer").Build();
    }
  }
}